src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs

```diff
@@ -39,6 +37,21 @@ internal SeparatedWithManyChildren(InternalSyntax.SyntaxList green, SyntaxNode?
 
                 return _children[i >> 1].Value;
             }
+
+            internal override int GetChildPosition(int index)
+            {
+                // If the previous sibling (ignoring separator) is not cached, but the next sibling
+                // (ignoring separator) is cached, use the next sibling to determine position.
+                int valueIndex = (index & 1) != 0 ? index - 1 : index;
+                if (valueIndex > 1
+                    && _children[(valueIndex - 2) >> 1].Value is null
+                    && (valueIndex >= Green.SlotCount - 2 || _children[(valueIndex + 2) >> 1].Value is { }))
+                {
+                    return GetChildPositionFromEnd(index);
```

Comment:
>GetChildPositionFromEnd [](http://example.com/codeflow?start=27&length=23)

It feels like the correlation between cache checks performed by ```GetChildPositionFromEnd``` and cache checks performed here would be more obvious if we were using ```GetCachedSlot``` helper rather than working with ```_children``` on the low level. Also, could we instead adjust implementation of ```GetChildPosition``` to perform the same cache tests for the following child? Then all the logic would be in one place, and, perhaps, we wouldn't need to provide a general ```GetChildPositionFromEnd``` helper, since it looks like we really want to handle just 3 specific situations.

It looks like the optimization targets very limited set of scenarios. For example, when we are far from the end and the next sibling isn't cached, but the one after the next is cached, we won't take advantage of that cached sibling. Would it make sense to generalize the optimization. For example, even if nothing is cached, it still might be faster to calculate from end simply because the item is closer to the end than to the front. Similarly, we could make decision based on the proximity of the first cached item going backwards vs. going forward. Perhaps we could maintain a bitmap of cached items to speed-up the process of finding closest cached items.

Reply:
> It looks like the optimization targets very limited set of scenarios.

Yes, the optimization specifically targets the scenario where the first iteration through the child list in reverse, so the previous siblings are not in the cache. That's the one scenario we have currently where there is a potential perf issue. The optimization here is simply to calculate the position based on the offset from the immediately following sibling when that sibling is in the cache; otherwise we use the existing approach.

We should limit the change here to fixing this one scenario, to avoid perf regressions in other cases. If additional scenarios arise, we can consider further optimizations.

Reply:
> Also, could we instead adjust implementation of `GetChildPosition` to perform the same cache tests for the following child? 

`GetChildPosition()` doesn't know about separated lists (where the separator is not in the cache) so it would be surprising for that method to skip over some siblings when determining whether using the following sibling is a better choice.

> Then all the logic would be in one place, and, perhaps, we wouldn't need to provide a general `GetChildPositionFromEnd` helper ...

I added `GetChildPositionFromEnd()` next to `GetChildPostion()` so the two loops were together. And I think we'd still need the cases that are handled in `GetChildPositionFromEnd()`, even if the methods were combined.

Reply:
> It feels like the correlation between cache checks performed by `GetChildPositionFromEnd` and cache checks performed here would be more obvious if we were using `GetCachedSlot` helper rather than working with _children on the low level. 

Updated.